### PR TITLE
Strings translations policy

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Why is this needed?
This adds explanations around our new support for translations in pro client.

### CODE REVIEW

File: dev-docs/explanations/string_changes_policy.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Policy on changing user-facing strings in Ubuntu Pro Client
+
+As of version 30 of `pro`, we support translations of user-facing strings throughout the program.
+These are maintained in `po` files in the `debian/po` directory of the git repository.
+Whenever a string is changed, added, or deleted, that has an impact on existing translations and
+therefore an impact on the user experience of `pro` in non-english locales.
+To minimize regressions in translations, we have the following policy in place regarding string changes.
+This policy is implemented in practice via a custom Github action that posts a questionare and prompt
+for justification on all Github PRs that modify the `messages` module of Ubuntu Pro Client.
```
Comment (s-makin) [inline dev-docs/explanations/string_changes_policy.md:9]:
```suggestion
for justification on all GitHub PRs that modify the `messages` module of Ubuntu Pro Client.
```

Comment (s-makin) [reply dev-docs/explanations/string_changes_policy.md:9]: I'm just gonna suggest a global search/replace for this :)

File: dev-docs/explanations/string_changes_policy.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+## Changing existing (already translated) strings
+
+All changes to existing strings will require justification in the Github PR that introduces the change.
+
+Acceptable justifications:
+- Typos: these will hopefully be minimal after adding spell-check CI, but still could occur.
```
Comment (s-makin) [inline dev-docs/explanations/string_changes_policy.md:16]: When is/was spell-check CI added? Not clear if this is already implemented or due to appear in the future

File: dev-docs/explanations/string_changes_policy.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  - Note: existing translations should be preserved when fixing typos
+- Factual errors: the nature of some features may change which will require us to update the description so that it is not wrong.
+- A Product Manager told us to change it: E.g. Product is renamed or description needs more info.
```
Comment (s-makin) [inline dev-docs/explanations/string_changes_policy.md:19]: I'd rephrase this, I'm not sure why (can't put my finger on it exactly) but the wording of this sounds a bit off. I'd suggest something more like:

Change requested by a Product Manager due to the product being renamed or service descriptions being updated.

File: dev-docs/explanations/string_changes_policy.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+When a change is made to a translated string, a best effort will be made to get Canonicalers to fix the translations of the changed string in the PR that introduces the change. However, a lack of complete translation replacements will not block the release of the changes.
```
Comment (s-makin) [inline dev-docs/explanations/string_changes_policy.md:21]:
```suggestion
When a change is made to a translated string, a best effort will be made to get Canonicalers to fix the translations of the changed string within the PR that introduces the change. However, a lack of complete translation replacements will not block the release of the changes.
```

Comment (s-makin) [reply dev-docs/explanations/string_changes_policy.md:21]: Should we point out that therefore we cannot guarantee that translations are always completely up to date?

Comment (s-makin) [reply dev-docs/explanations/string_changes_policy.md:21]: Maybe something more like:

When a change is made to a translated string, we (Canonical) will endeavour to provide translations for the changed string within the PR that introduces the change. However, these translations are provided on a "best effort" basis and we will not block the release of the changes due to incomplete translations.

File: dev-docs/explanations/translations.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
+# Translations
+
+Ubuntu Pro Client is a special package in Ubuntu: it is updated frequently in all Ubuntu LTS releases.
+That makes language packs unsuitable for providing and updating translations for Ubuntu Pro Client.
+
+In addition, Ubuntu Pro Client includes lots of language with potential commercial and contractual
+implications related to Canonical's Ubuntu Pro offering. That means we need to maintain tight control
+over the language used in all messages.
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:8]: "We need to maintain tight control" sounds a bit "off". I'd suggest something more like:

That means Canonical must be both careful and specific about the language used in all messages to ensure that we do not provide misleading or incorrect information.

File: dev-docs/explanations/translations.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+## How are translations delivered and updated?
+
+Normally, packages in `main` are translated via language packs; however language packs are only updated
+in stable releases around the time of point releases. If Ubuntu Pro translations were put in the language pack,
+then the language pack wouldn’t get updated as we continue to SRU new Ubuntu Pro Client versions back to
+old Ubuntu releases. That would cause the translations provided by the language pack to drift out of sync with
+the version of Ubuntu Pro Client installed.
+
+Because of this, Ubuntu Pro Client is not included in the Ubuntu language pack and instead ships all translations
+directly in our own source package. However, translations are kept in a separate binary package so that it is not
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:19]: "it is not required" - unclear what "it" refers to in this part

File: dev-docs/explanations/translations.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+In order to ship translation files (`.mo` files) in the `ubuntu-pro-client-l10n` binary package, this package is
+explicitly excluded from language pack translation stripping done by the `pkgbinarymangler` package
+(see [LP: #2037584](https://bugs.launchpad.net/ubuntu/+source/pkgbinarymangler/+bug/2037584)).
+
+## How are translations maintained and added?
+
+Translations live as `.po` files in the `debian/po/` directory.
+
+Because many of the pro-client’s messages are commercial in nature, a bad translation could potentially have odd or
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:32]:
```suggestion
Because many of the Pro Client’s messages are commercial in nature, a bad translation could potentially have odd or
```

File: dev-docs/explanations/translations.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+misleading contractual implications. Because of that, we generally want all translations to come from Canonical employees.
+
+Regarding contributions, we will:
+- Accept Canonical-employee PRs of translations in Github.
+- Consider community bugs/issues regarding translation wording/suggestions
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:37]:
```suggestion
- Consider community bugs/issues regarding translation wording/suggestions.
```

File: dev-docs/explanations/translations.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  - We’ll get input from a Canonical employee who speaks the language in question.
+- **Not** accept community PRs directly changing translations
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:39]:
```suggestion
- **Not** accept community PRs directly changing translations.
```

File: dev-docs/explanations/translations.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+We generally will not support translations to a language that no Canonical-employee understands and is willing to review/help maintain.
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:41]: Wording is a bit confusing, I had to read it twice. Suggest something more like:

For a language to be supported, there must be at least one Canonical employee who speaks/reads/understands that language and is willing to review/help maintain the translation.

File: dev-docs/explanations/translations.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+We will set up a project on https://translations.launchpad.net where only Canonical-employees have permission to contribute
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:43]:
```suggestion
We will set up a project on https://translations.launchpad.net where only Canonical employees have permission to contribute
```

File: dev-docs/explanations/translations.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+translations to Ubuntu Pro Client. Before every release of Ubuntu Pro Client, we will download the latest translations
+from this launchpad project and integrate them. *Note that this is not done yet.*
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:45]:
```suggestion
from this Launchpad project and integrate them. *Note that this is not done yet.*
```

File: dev-docs/explanations/translations.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+# What parts of pro-client will get translated?
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:47]:
```suggestion
# What parts of the Pro Client will get translated?
```

File: dev-docs/explanations/translations.md [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+These **do** get translated:
+- Human-readable output of all `pro` cli commands
+  - Including:
+    - help text (called out because this is handled in a special way in pro-client)
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:52]:
```suggestion
    - help text (called out because this is handled in a special way in Pro Client)
```

File: dev-docs/explanations/translations.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    - `pro status` table values including: “enabled”, ”disabled”, “available”, etc
+- Pro-related messages inserted in apt upgrade
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:54]:
```suggestion
- Pro-related messages inserted into `apt upgrade`
```

File: dev-docs/explanations/translations.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+- Messages inserted in MOTD
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:55]:
```suggestion
- Messages inserted into the MOTD
```

File: dev-docs/explanations/translations.md [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+These do **not** get translated:
+- JSON output of `pro` cli commands
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:58]:
```suggestion
- JSON output of `pro` CLI commands
```

File: dev-docs/explanations/translations.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  - For example: {“status”: “enabled”} – both “status” and “enabled” should remain english
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:59]:
```suggestion
  - For example: {“status”: “enabled”} – both “status” and “enabled” should remain in English
```

File: dev-docs/explanations/translations.md [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  - Exception: some values in the json are messages intended for humans – these should get translated
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:60]:
```suggestion
  - Exception: some values in the JSON are messages intended for humans – these should get translated
```

File: dev-docs/explanations/translations.md [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+- Cli flags
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:61]:
```suggestion
- CLI flags
```

**Review by s-makin — COMMENTED** [review]
Just a few suggestions/nits

File: dev-docs/explanations/translations.md (hunk 12 above)
Comment (orndorffgrant) [reply dev-docs/explanations/translations.md:45]: Yes - this part isn't done, but I don't want it to block publishing this doc. I'm open to different ways of communicating that though.

File: dev-docs/explanations/string_changes_policy.md (hunk 1 above)
Comment (orndorffgrant) [reply dev-docs/explanations/string_changes_policy.md]: Done :)

File: dev-docs/explanations/string_changes_policy.md (hunk 4 above)
Comment (orndorffgrant) [reply dev-docs/explanations/string_changes_policy.md:21]: That sounds much better, I'll change it tthat

File: dev-docs/explanations/string_changes_policy.md (hunk 2 above)
Comment (orndorffgrant) [reply dev-docs/explanations/string_changes_policy.md:16]: It will be added by the time this goes live :), I'll clarify that in the text

File: dev-docs/explanations/translations.md (hunk 6 above)
Comment (orndorffgrant) [reply dev-docs/explanations/translations.md:19]: Updating it to be clearer

**Comment by orndorffgrant** [comment]
@s-makin Thank you for the review! I've accepted all of your suggested changes and attempted to clarify the confusing parts you called out. Let me know what you think

File: dev-docs/explanations/translations.md (hunk 12 above)
Comment (s-makin) [reply dev-docs/explanations/translations.md]: That's fine, I just wanted to check that this wasn't a placeholder that you'd forgotten :)

File: dev-docs/explanations/translations.md [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+Normally, packages in `main` are translated via language packs; however language packs are only updated
+in stable releases around the time of point releases. If Ubuntu Pro translations were put in the language pack,
+then the language pack wouldn’t get updated as we continue to SRU new Ubuntu Pro Client versions back to
+old Ubuntu releases. That would cause the translations provided by the language pack to drift out of sync with
+the version of Ubuntu Pro Client installed.
+
+Because of this, Ubuntu Pro Client is not included in the Ubuntu language pack and instead ships all translations
+directly in our own source package. However, translations are kept in a separate binary package so that the translation
+fils are not required to be installed with Ubuntu Pro Client (and therefore not required in `ubuntu-minimal`).
```
Comment (s-makin) [inline dev-docs/explanations/translations.md:21]:
```suggestion
files are not required to be installed with Ubuntu Pro Client (and therefore not required in `ubuntu-minimal`).
```

**Review by s-makin — APPROVED** [review]
LGTM :) there's just one little nit in one of the changes, but otherwise I'm happy if you're happy

File: dev-docs/explanations/translations.md (hunk 21 above)
Comment (orndorffgrant) [reply dev-docs/explanations/translations.md:21]: :upside_down_face: I'll fix it

File: dev-docs/howtoguides/spellcheck.md [hunk 22]
```diff
@@ -0,0 +1,20 @@
+# How to spellcheck messages
+
+We have CI that runs on every PR which looks for spelling errors in our messages module, but it is often useful to run it locally as well.
```
Comment (s-makin) [inline dev-docs/howtoguides/spellcheck.md:3]: What language(s) does this check?

File: dev-docs/howtoguides/spellcheck.md [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+
+Then run it on our messages module using our list of extra allowed words.
+
+```
+hunspell -p ./tools/spellcheck-allowed-words.txt -l ./uaclient/messages/__init__.py
+```
+
+That particular `hunspell` command will print any incorrectly spelled words to stdout.
+
+`hunspell` has other options including an interactive mode for fixing errors. Check out the [`hunspell` documentation(https://github.com/hunspell/hunspell#documentation) for more information.
```
Comment (s-makin) [inline dev-docs/howtoguides/spellcheck.md:18]:
```suggestion
`hunspell` has other options including an interactive mode for fixing errors. Check out the [`hunspell` documentation](https://github.com/hunspell/hunspell#documentation) for more information.
```

**Review by s-makin — COMMENTED** [review]
Generally fine with the new instructions - just a couple more nits and then gtg :)

File: dev-docs/howtoguides/spellcheck.md (hunk 22 above)
Comment (orndorffgrant) [reply dev-docs/howtoguides/spellcheck.md:3]: Only english for now.

@s-makin is this better?
```suggestion
We have CI that runs on every PR which looks for spelling errors in our English messages, but it is often useful to run it locally as well.
```

Comment (s-makin) [reply dev-docs/howtoguides/spellcheck.md:3]: Fine by me - just so anyone working on translations doesn't waste their time running it :)

**Comment by orndorffgrant** [comment]
Thank you again @s-makin! Just resolved the last couple nits

**Comment by orndorffgrant** [comment]
@renanrodrigo or @lucasmoura do you want to review before we merge?

**Comment by renanrodrigo** [comment]
I can take a quick look sure

**Review by renanrodrigo — APPROVED** [review]
